File: examples/simulation/ur/cpp/cartesian_motion_controller.cpp
#include <Eigen/Dense>
#include <fstream>
#include <iostream>
#include <sdu_controllers/controllers/operational_space_controller.hpp>
#include <sdu_controllers/math/forward_dynamics.hpp>
#include <sdu_controllers/math/inverse_dynamics_joint_space.hpp>
#include <sdu_controllers/math/trajectory_generation.hpp>
#include <sdu_controllers/kinematics/forward_kinematics.hpp>
#include <sdu_controllers/models/ur_robot_model.hpp>
#include <sdu_controllers/safety/safety_verifier.hpp>
#include <sdu_controllers/utils/utility.hpp>

using namespace csv;
using namespace Eigen;
using namespace sdu_controllers;
using namespace sdu_controllers::utils;

constexpr double pi = 3.14159265358979323846;

int main()
{
  // Setup writing of output trajectory to csv.
  std::ofstream output_filestream;
  output_filestream.open("output_cartesian.csv");
  auto csv_writer = make_csv_writer(output_filestream);

  // Initialize robot model and parameters
  auto robot_model = std::make_shared<models::URRobotModel>(models::URRobotModel::RobotType::ur5e);
  double freq = 500.0;
  double dt = 1.0 / freq;
  const double total_t = 4.0;   // seconds (two full circles)
  const size_t steps = static_cast<size_t>(total_t * freq);

  // Circle parameters
  const double radius = 0.05;               // 5 cm
  const double circle_freq = 0.5;            // Hz
  const double omega = 2.0 * pi * circle_freq;
  const double ramp_time = 0.3;              // s

  double Kp_pos_value = 16250.0;
  double Kp_orient_value = 16250.0;
  double Kd_pos_value = 200.0;
  double Kd_orient_value = 3.0;
  double N_value = 1;
  uint16_t ROBOT_DOF = robot_model->get_dof();

  VectorXd Kp_pos_vec = VectorXd::Ones(3) * Kp_pos_value;
  VectorXd Kp_orient_vec = VectorXd::Ones(3) * Kp_orient_value;
  VectorXd Kd_pos_vec = VectorXd::Ones(3) * Kd_pos_value;
  VectorXd Kd_orient_vec = VectorXd::Ones(3) * Kd_orient_value;
  VectorXd N_vec = VectorXd::Ones(6) * N_value;

  MatrixXd Kp = MatrixXd::Zero(6, 6);
  Kp.setIdentity();
  Kp.block<3, 3>(0,0) = Kp_pos_vec.asDiagonal();
  Kp.block<3, 3>(3,3) = Kp_orient_vec.asDiagonal();
  MatrixXd Kd = MatrixXd::Zero(6, 6);
  Kd.setIdentity();
  Kd.block<3, 3>(0,0) = Kd_pos_vec.asDiagonal();
  Kd.block<3, 3>(3,3) = Kd_orient_vec.asDiagonal();

  // ZYZ mode (default): x_d is [position(3), ZYZ angles(3)].
  controllers::OperationalSpaceController osc_controller(Kp, Kd, robot_model);

  VectorXd x_d(6);
  VectorXd dx_d(6);
  VectorXd ddx_d(6);

  VectorXd q(ROBOT_DOF);
  VectorXd dq(ROBOT_DOF);
  Vector<double, 6> he = VectorXd::Zero(6);
  q << 0.0, -1.5707, -1.5707, -1.5707, 1.5707, 0.0;
  dq << 0.0, 0.0, 0.0, 0.0, 0.0, 0.0;

  // Circle centre and fixed orientation derived from the initial end-effector pose.
  Matrix4d T0 = robot_model->get_fk_solver().forward_kinematics(q);
  Vector3d center = T0.block<3, 1>(0, 3);
  center[0] -= radius;   // shift so the trajectory starts at the robot's initial position
  Vector3d rot_zyz_d = T0.topLeftCorner<3, 3>().eulerAngles(2, 1, 2);

  // Control loop
  for (size_t step = 0; step < steps; ++step)
  {
    const double t = step * dt;

    // Desired Cartesian pose, velocity and acceleration
    Vector3d pos_d, dpos_d, ddpos_d;
    math::circular_trajectory(center, radius, omega, t, ramp_time, pos_d, dpos_d, ddpos_d);

    x_d << pos_d, rot_zyz_d;
    dx_d << dpos_d, Vector3d::Zero();
    ddx_d << ddpos_d, Vector3d::Zero();

    // Add noise to q and dq
    VectorXd q_meas = q;
    VectorXd dq_meas = dq;
    //add_noise_to_vector(q_meas, 0.0, 0.001);
    //add_noise_to_vector(dq_meas, 0.0, 0.001);

    // Controller
    osc_controller.step(x_d, dx_d, ddx_d, q_meas, dq_meas);
    VectorXd y = osc_controller.get_output();
    VectorXd tau = robot_model->inverse_dynamics(q_meas, dq_meas, y, he);

    // Simulation
    VectorXd ddq = robot_model->forward_dynamics(q, dq, tau);
    // integrate to get velocity
    dq += ddq * dt;
    // integrate to get position
    q += dq * dt;

    MatrixXd T = robot_model->get_fk_solver().forward_kinematics(q);
    VectorXd pos = T.block<3, 1>(0, 3);
    Matrix3d rot_mat = T.block<3,3>(0, 0);
    Vector3d rpy_zyz = rot_mat.eulerAngles(2, 1, 2); // ZYZ representation
    VectorXd temp(q.size()+pos.size()+rpy_zyz.size());
    temp << q, pos, rpy_zyz;
    csv_writer << eigen_to_std_vector(temp);
  }
  output_filestream.close();
  std::cout << "Simulation complete. Output written to output_cartesian.csv" << std::endl;
}

File: examples/simulation/ur/python/cartesian_motion_controller.py
"""Operational-space controller simulation example - UR5e following a Cartesian circle.

The robot tracks a circular reference trajectory in the XY plane using the
operational-space (inverse-dynamics) controller in ZYZ orientation mode.
"""

import numpy as np
import sdu_controllers
from sdu_controllers.common import circular_trajectory

PI = 3.14159265358979323846


def rotation_to_zyz(rot: np.ndarray) -> np.ndarray:
    """Extract ZYZ Euler angles from a 3x3 rotation matrix (matches Eigen convention)."""
    theta = np.arccos(np.clip(rot[2, 2], -1.0, 1.0))
    if abs(np.sin(theta)) < 1e-10:
        phi = 0.0
        psi = np.arctan2(-rot[0, 1], rot[0, 0]) if rot[2, 2] > 0 else np.arctan2(rot[0, 1], -rot[0, 0])
    else:
        phi = np.arctan2(rot[2, 1], rot[2, 0])
        psi = np.arctan2(rot[1, 2], -rot[0, 2])
    return np.array([phi, theta, psi])


def main():
    freq = 500.0
    dt = 1.0 / freq
    total_t = 4.0  # seconds (two full circles)
    steps = int(total_t * freq)

    # Circle parameters
    radius = 0.05
    circle_freq = 0.5  # Hz
    omega = 2.0 * PI * circle_freq
    ramp_time = 0.3

    robot_model = sdu_controllers.models.URRobotModel(sdu_controllers.models.RobotType.ur5e)
    dof = robot_model.get_dof()

    Kp_pos, Kp_orient = 16250.0, 16250.0
    Kd_pos, Kd_orient = 200.0, 3.0

    Kp = np.eye(6)
    Kp[0:3, 0:3] = np.eye(3) * Kp_pos
    Kp[3:6, 3:6] = np.eye(3) * Kp_orient
    Kd = np.eye(6)
    Kd[0:3, 0:3] = np.eye(3) * Kd_pos
    Kd[3:6, 3:6] = np.eye(3) * Kd_orient

    # ZYZ mode (default): x_d is [position(3), ZYZ angles(3)].
    controller = sdu_controllers.controllers.OperationalSpaceController(Kp, Kd, robot_model)

    q = np.array([0.0, -1.5707, -1.5707, -1.5707, 1.5707, 0.0])
    dq = np.zeros(dof)
    he = np.zeros(6)

    T0 = robot_model.get_fk_solver().forward_kinematics(q)
    center = T0[:3, 3].copy()
    center[0] -= radius
    rot_zyz_d = rotation_to_zyz(T0[:3, :3])

    for step in range(steps):
        t = step * dt
        pos_d, dpos_d, ddpos_d = circular_trajectory(center, radius, omega, t, ramp_time)

        x_d = np.concatenate([pos_d, rot_zyz_d])
        dx_d = np.concatenate([dpos_d, np.zeros(3)])
        ddx_d = np.concatenate([ddpos_d, np.zeros(3)])

        controller.step(x_d, dx_d, ddx_d, q, dq)
        y = controller.get_output()

        tau = robot_model.inverse_dynamics(q, dq, y, he)
        ddq = robot_model.forward_dynamics(q, dq, tau)
        dq += ddq * dt
        q += dq * dt

    print("Simulation complete.")


if __name__ == "__main__":
    main()
